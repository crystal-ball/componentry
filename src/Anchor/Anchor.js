import elem from '../elem-factory'
import { actionClasses } from '../utils/componentry'
import { useTheme } from '../Theme/Theme'

/**
 * [Anchor component üìù](https://componentry.design/components/anchor)
 */
export default function Anchor(props) {
  const { variant = 'a', ...merged } = {
    ...useTheme('Anchor'),
    ...props,
  }

  return elem({
    as: 'a',
    elemClassName: actionClasses(variant, merged),
    ...merged,
  })
}
Anchor.displayName = 'Anchor'
