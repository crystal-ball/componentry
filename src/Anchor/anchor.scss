/* componentry/src/Anchor/anchor */

$a: () !default;

$a-primary: () !default;

$anchor-button-overrides: () !default;

// --------------------------------------------------------
// Normalizing buttons styled as anchors

// These overrides apply to Button instances using an a base class, they
// normalize browser button styles to match anchor styles

@include style-rule-from-map(
  'button.a',
  map-assign(
    $anchor-button-overrides,
    (
      border: none,
      background-color: transparent,
      cursor: pointer,
      font: inherit,
      padding: 0,
      text-align: inherit,
    )
  )
);

// --------------------------------------------------------
// Base class

@include style-rule-from-map('.a', map-assign($a, ()));

// --------------------------------------------------------
// Primary variant

$a-primary-base: (
  color: get($palette, 'primary.300'),
  '&:hover': (
    color: get($palette, 'primary.500'),
  ),
);

@each $color in theme-get('palette.theme-colors') {
  $a-primary-base: map-merge(
    $a-primary-base,
    (
      '&.a-#{$color}-color': (
        color: theme-get('palette.#{$color}.300'),
        '&:hover': (
          color: theme-get('palette.#{$color}.400'),
        ),
      ),
    )
  );
}

@include style-rule-from-map('.a-primary', map-assign($a-primary-base, $a-primary));

// --- ℹ️ Disabled anchor styles
// Componentry doesn't include disabled anchor styles as a UX best
// practice, but if necessary it's best to define normal anchor styles
// using `a[href]` as the selector, and `a` as a 'disabled' anchor and
// then don't pass an href to create a 'placeholder link'
