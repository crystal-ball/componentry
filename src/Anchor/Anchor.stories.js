/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'

import Anchor from './Anchor'
import Block from '../Block/Block'

const colors = {
  Primary: 'primary',
  Secondary: 'secondary',
  Info: 'info',
  Light: 'light',
  Dark: 'dark',
  Success: 'success',
  Warning: 'warning',
  Danger: 'danger',
}

storiesOf('Anchor', module)
  .add('<Anchor />', () => <Anchor href='#'>Componentry anchor</Anchor>)
  .add('Collection', () => (
    <Block>
      <h2 className='storybook-section-header'>Anchor</h2>
      {Object.keys(colors).map(color => (
        <p key={color}>
          <Anchor color={colors[color]} href='#'>
            {color}
          </Anchor>
        </p>
      ))}

      <h2 className='storybook-section-header'>Disabled</h2>
      <div className='btn-container-x'>
        <Anchor disabled href='#'>
          Disabled
        </Anchor>
      </div>

      <h2 className='storybook-section-header'>Button variant anchor</h2>
      <div className='btn-container-x'>
        {Object.keys(colors).map(color => (
          <Anchor button color={colors[color]} href='#' key={color}>
            {color}
          </Anchor>
        ))}
      </div>

      <h2 className='storybook-section-header'>Outline button variant anchor</h2>
      <div className='btn-container-x'>
        {Object.keys(colors).map(color => (
          <Anchor button outline={colors[color]} href='#' key={color}>
            {color}
          </Anchor>
        ))}
      </div>
    </Block>
  ))
