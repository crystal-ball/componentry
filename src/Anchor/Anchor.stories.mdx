import { Meta, Title, Props } from '@storybook/addon-docs/blocks'
import { boolean, select } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import Anchor from './Anchor'
import { ElementsProvider, colors, size } from '../story-resources'

<Meta title='Components|Anchor' component={Anchor} />

<ElementsProvider>

# Anchors

Anchor element with an href:

<Anchor href='#'>Anchor element</Anchor>

<Props of={Anchor} />

## Theme colors

Pass a `color` to set a theme color background:

<div className='d-flex justify-content-center flex-column'>
  {colors.map(color => (
    <Anchor color={color} href='#' key={color}>
      Achor element
    </Anchor>
  ))}
</div>

## Disabled

Pass `disabled` to make an anchor disabled:

<Anchor disabled>Anchor element</Anchor>

## Button class

Create a button element with anchor styles by passing `as='button'` to an Anchor:

<div className='d-flex justify-content-around btn-container-x'>
  <Anchor href='#'>Anchor element</Anchor>
  <Anchor as='button' onClick={action('button anchor')}>
    Button element
  </Anchor>
</div>

</ElementsProvider>
