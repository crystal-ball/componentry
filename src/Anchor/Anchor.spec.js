import React from 'react'
import { render } from '@testing-library/react'

import Anchor from './Anchor'
import elementTests from '../../test/element-tests'

describe('<Anchor/>', () => {
  // Basic library element test suite
  elementTests(Anchor)

  test('should render type classes for anchor', () => {
    const { getByText } = render(
      <Anchor color='success' size='sm'>
        Link
      </Anchor>,
    )
    expect(getByText('Link')).toHaveClass('a a-success a-sm')
  })

  test('When used to create a button styled anchor, then btn is rendered', () => {
    const { getByText } = render(<Anchor variant='btn'>Link</Anchor>)
    expect(getByText('Link')).toHaveClass('btn')
    expect(getByText('Link')).not.toHaveClass('a')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Anchor /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(<Anchor>Link</Anchor>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders btn variants correctly', () => {
    const { container } = render(
      <>
        <Anchor variant='btn'>Link</Anchor>
        <Anchor variant='btn' outline='success' size='sm'>
          Link
        </Anchor>
      </>,
    )
    expect(container).toMatchSnapshot()
  })
})
