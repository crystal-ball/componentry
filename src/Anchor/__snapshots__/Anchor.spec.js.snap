// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anchor /> snapshots renders btn variants correctly 1`] = `
<div>
  <a
    class="btn"
  >
    Link
  </a>
  <a
    class="btn btn-outline-success btn-sm"
  >
    Link
  </a>
</div>
`;

exports[`<Anchor /> snapshots renders correctly 1`] = `
<a
  class="a"
>
  Link
</a>
`;
